File: src/main/java/com/flink/tutorials/java/chapter5_time/CoGroupExample.java
package com.flink.tutorials.java.chapter5_time;

import org.apache.flink.api.common.functions.CoGroupFunction;
import org.apache.flink.api.common.typeinfo.Types;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.TimeCharacteristic;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.windowing.assigners.TumblingProcessingTimeWindows;
import org.apache.flink.streaming.api.windowing.time.Time;
import org.apache.flink.util.Collector;

public class CoGroupExample {

    public static void main(String[] args) throws Exception {

        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        // 使用EventTime时间语义
        env.setStreamTimeCharacteristic(TimeCharacteristic.ProcessingTime);

        DataStream<String> socketSource1 = env.socketTextStream("localhost", 9000);
        DataStream<String> socketSource2 = env.socketTextStream("localhost", 9001);

        DataStream<Tuple2<String, Integer>> input1 = socketSource1.map(
                line -> {
                    String[] arr = line.split(" ");
                    String id = arr[0];
                    int t = Integer.parseInt(arr[1]);
                    return Tuple2.of(id, t);
                })
                .returns(Types.TUPLE(Types.STRING, Types.INT));
        DataStream<Tuple2<String, Integer>> input2 = socketSource2.map(
                line -> {
                    String[] arr = line.split(" ");
                    String id = arr[0];
                    int t = Integer.parseInt(arr[1]);
                    return Tuple2.of(id, t);
                })
                .returns(Types.TUPLE(Types.STRING, Types.INT));

        DataStream<String> coGroupResult = input1.coGroup(input2)
                .where(i1 -> i1.f0)
                .equalTo(i2 -> i2.f0)
                .window(TumblingProcessingTimeWindows.of(Time.seconds(10)))
                .apply(new MyCoGroupFunction());

        coGroupResult.print();

        env.execute("window cogroup function");
    }

    public static class MyCoGroupFunction implements CoGroupFunction<Tuple2<String, Integer>, Tuple2<String, Integer>, String> {
        @Override
        public void coGroup(Iterable<Tuple2<String, Integer>> input1, Iterable<Tuple2<String, Integer>> input2, Collector<String> out) {
            input1.forEach(element -> System.out.println("input1 :" + element.f1));
            input2.forEach(element -> System.out.println("input2 :" + element.f1));
        }
    }
}

File: src/main/scala/com/flink/tutorials/scala/chapter5_time/CoGroupExample.scala
package com.flink.tutorials.scala.api.chapter5

import java.lang

import org.apache.flink.api.common.functions.CoGroupFunction
import org.apache.flink.streaming.api.TimeCharacteristic
import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.api.windowing.assigners.TumblingProcessingTimeWindows
import org.apache.flink.streaming.api.windowing.time.Time
import org.apache.flink.util.Collector

import scala.collection.JavaConverters._

object CoGroupExample {

  def main(args: Array[String]): Unit = {

    val senv = StreamExecutionEnvironment.getExecutionEnvironment
    senv.setStreamTimeCharacteristic(TimeCharacteristic.ProcessingTime)

    val socketSource1 = senv.socketTextStream("localhost", 9000)
    val socketSource2 = senv.socketTextStream("localhost", 9001)

    val input1: DataStream[(String, Int)] = socketSource1.flatMap {
      (line: String, out: Collector[(String, Int)]) => {
        val array = line.split(" ")
        if (array.size == 2) {
          out.collect((array(0), array(1).toInt))
        }
      }
    }

    val input2: DataStream[(String, Int)] = socketSource2.flatMap {
      (line: String, out: Collector[(String, Int)]) => {
        val array = line.split(" ")
        if (array.size == 2) {
          out.collect((array(0), array(1).toInt))
        }
      }
    }


    val coGroupResult = input1.coGroup(input2)
      .where(i1 => i1._1)
      .equalTo(i2 => i2._1)
      .window(TumblingProcessingTimeWindows.of(Time.seconds(60)))
      .apply(new MyCoGroupFunction)

    coGroupResult.print()

    senv.execute("window cogroup function")
  }

  class MyCoGroupFunction extends CoGroupFunction[(String, Int), (String, Int), String] {

    // 这里的类型是Java的Iterable，需要引用 collection.JavaConverters._ 并转成Scala
    override def coGroup(input1: lang.Iterable[(String, Int)], input2: lang.Iterable[(String, Int)], out: Collector[String]): Unit = {
      input1.asScala.foreach(element => out.collect("input1 :" + element.toString()))
      input2.asScala.foreach(element => out.collect("input2 :" + element.toString()))
    }
  }

}
